go/vt/vtgate/planbuilder/dml.go

```diff
@@ -36,34 +36,45 @@ func dmlFormatter(buf *sqlparser.TrackedBuffer, node sqlparser.SQLNode) {
 
 // buildUpdatePlan builds the instructions for an UPDATE statement.
 func buildUpdatePlan(upd *sqlparser.Update, vschema VSchema) (*engine.Route, error) {
-	route := &engine.Route{
+	er := &engine.Route{
 		Query: generateQuery(upd),
 	}
-	updateTable, _ := upd.Table.Expr.(sqlparser.TableName)
-
-	var err error
-	route.Table, err = vschema.Find(updateTable)
+	bldr, err := processTableExprs(upd.TableExprs, vschema)
 	if err != nil {
 		return nil, err
 	}
-	route.Keyspace = route.Table.Keyspace
+	rb, ok := bldr.(*route)
+	if !ok {
+		return nil, errors.New("unsupported: multi-table update statement in sharded keyspace")
+	}
 	if hasSubquery(upd) {
 		return nil, errors.New("unsupported: subqueries in DML")
 	}
-	if !route.Keyspace.Sharded {
-		route.Opcode = engine.UpdateUnsharded
-		return route, nil
+	er.Keyspace = rb.ERoute.Keyspace
+	if !er.Keyspace.Sharded {
+		er.Opcode = engine.UpdateUnsharded
+		return er, nil
 	}
-
-	err = getDMLRouting(upd.Where, route)
+	if len(rb.Symtab().tables) != 1 {
+		return nil, errors.New("unsupported: multi-table update statement in sharded keyspace")
+	}
+	var tableName sqlparser.TableName
+	for t := range rb.Symtab().tables {
+		tableName = t
+	}
+	er.Table, err = vschema.Find(tableName)
```

Comment:
This sucks :). We can (later) fix this by adding a vindexTable field to the table in symtab.